### PR TITLE
browser: set up extension protocol handler for each browser context

### LINKED ISSUE #4991: Custom Protocol Does Not Work With Different WebView Partition
* Electron version: 0.37.3
* Operating system: Windows 8.1

When using `electron.protocol.registerHttpProtocol` it works fine using the default `partition` on `<webview>`. But when using a different partition name it does not.

It appears this was handled in #4551 only to be reverted in #4588. I do not see an issue tracking the effort.

### LINKED ISSUE #5896: Crash loading devtools extension when using partition
If you specify a partition in your `BrowserWindow` options then opening the devtools with a third-party extension registered causes the app to crash.

Failing spec is available in https://github.com/electron/electron/blob/partition-devtools-crash/spec/api-browser-window-spec.js#L884-L901

```
Process:               Electron [60075]
Path:                  /Users/USER/*/Electron.app/Contents/MacOS/Electron
Identifier:            com.github.electron
Version:               1.2.1 (1.2.1)
Code Type:             X86-64 (Native)
Parent Process:        Python [60074]
Responsible:           Terminal [494]
User ID:               501

Date/Time:             2016-06-06 16:43:34.407 -0700
OS Version:            Mac OS X 10.11.5 (15F34)
Report Version:        11
Anonymous UUID:        9E9CB434-A4C0-CA4D-716E-2DA7D958B49C

Sleep/Wake UUID:       587C2E41-329E-448B-8575-B34940111114

Time Awake Since Boot: 88000 seconds
Time Since Wake:       13000 seconds

System Integrity Protection: enabled

Crashed Thread:        0  CrBrowserMain  Dispatch queue: com.apple.main-thread

Exception Type:        EXC_BAD_ACCESS (SIGSEGV)
Exception Codes:       KERN_INVALID_ADDRESS at 0x0000000000000008

VM Regions Near 0x8:
--> 
    __TEXT                 000000010d893000-000000010d894000 [    4K] r-x/rwx SM=COW  /Users/USER/*/Electron.app/Contents/MacOS/Electron

Thread 0 Crashed:: CrBrowserMain  Dispatch queue: com.apple.main-thread
0   com.github.electron.framework 	0x000000010da7f101 atom::WebContentsPermissionHelper::RequestPermission(content::PermissionType, base::Callback<void (bool), (base::internal::CopyMode)1> const&, bool) + 33 (web_contents_permission_helper.cc:58)
1   com.github.electron.framework 	0x000000010da7f63b atom::WebContentsPermissionHelper::RequestOpenExternalPermission(base::Callback<void (bool), (base::internal::CopyMode)1> const&, bool) + 59 (web_contents_permission_helper.cc:100)
2   com.github.electron.framework 	0x000000010d9eaebe atom::(anonymous namespace)::HandleExternalProtocolInUI(GURL const&, base::Callback<content::WebContents* (), (base::internal::CopyMode)1> const&, bool) + 206 (atom_resource_dispatcher_host_delegate.cc:38)
3   com.github.electron.framework 	0x000000010d9ebb6c void base::internal::RunnableAdapter<void (*)(GURL const&, base::Callback<content::WebContents* (), (base::internal::CopyMode)1> const&, bool)>::Run<GURL const&, base::Callback<content::WebContents* (), (base::internal::CopyMode)1> const&, bool const&>(GURL const&&&, base::Callback<content::WebContents* (), (base::internal::CopyMode)1> const&&&, bool const&&&) + 76 (bind_internal.h:159)
4   com.github.electron.framework 	0x000000010d9ebb08 void base::internal::InvokeHelper<false, void, base::internal::RunnableAdapter<void (*)(GURL const&, base::Callback<content::WebContents* (), (base::internal::CopyMode)1> const&, bool)> >::MakeItSo<GURL const&, base::Callback<content::WebContents* (), (base::internal::CopyMode)1> const&, bool const&>(base::internal::RunnableAdapter<void (*)(GURL const&, base::Callback<content::WebContents* (), (base::internal::CopyMode)1> const&, bool)>, GURL const&&&, base::Callback<content::WebContents* (), (base::internal::CopyMode)1> const&&&, bool const&&&) + 72 (bind_internal.h:312)
5   com.github.electron.framework 	0x000000010d9ebaae base::internal::Invoker<base::IndexSequence<0ul, 1ul, 2ul>, base::internal::BindState<base::internal::RunnableAdapter<void (*)(GURL const&, base::Callback<content::WebContents* (), (base::internal::CopyMode)1> const&, bool)>, void (GURL const&, base::Callback<content::WebContents* (), (base::internal::CopyMode)1> const&, bool), GURL const&, base::Callback<content::WebContents* (), (base::internal::CopyMode)1> const&, bool&>, base::internal::InvokeHelper<false, void, base::internal::RunnableAdapter<void (*)(GURL const&, base::Callback<content::WebContents* (), (base::internal::CopyMode)1> const&, bool)> >, void ()>::Run(base::internal::BindStateBase*) + 158 (bind_internal.h:362)
6   libbase.dylib                 	0x000000010fec126b base::debug::TaskAnnotator::RunTask(char const*, base::PendingTask const&) + 187
```
Refs https://github.com/electron/devtron/issues/75

### LINKED ISSUE #5925: Crash in Electron 1.2.1 when closing window (AskForOptions)
I'm trying to narrow down the cause of this crash, but it happens about 50% of the time when users of N1 exit the "welcome" flow and the app tries to close one window and open two others. It seems to be failing deep within `atom::internal::AskForOptions`.

Would love any input or advice to narrow this down.

Two (slightly different) logs so far:

```
Thread 0 Crashed:: CrBrowserMain  Dispatch queue: com.apple.main-thread
0   libsystem_platform.dylib      	0x00007fff8f001551 _platform_memmove$VARIANT$Ivybridge + 49
1   libc++.1.dylib                	0x00007fff8a6ea225 std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >::__init(char const*, unsigned long) + 91
2   com.github.electron.framework 	0x00000001044d9222 0x1042c0000 + 2200098
3   com.github.electron.framework 	0x00000001043a78de mate::Converter<net::URLRequest const*, void>::ToV8(v8::Isolate*, net::URLRequest const*) + 350
4   com.github.electron.framework 	0x00000001042cdb6c v8::Local<v8::Value> mate::ConvertToV8<net::URLRequest const*>(v8::Isolate*, net::URLRequest const* const&) + 12
5   com.github.electron.framework 	0x00000001042fd45c mate::internal::V8FunctionInvoker<void (net::URLRequest const*, v8::Local<v8::Value>)>::Go(v8::Isolate*, mate::internal::SafeV8Function const&, net::URLRequest const*, v8::Local<v8::Value>) + 172
6   com.github.electron.framework 	0x00000001042fd6e9 void base::internal::RunnableAdapter<void (*)(v8::Isolate*, mate::internal::SafeV8Function const&, net::URLRequest const*, v8::Local<v8::Value>)>::Run<v8::Isolate* const&, mate::internal::SafeV8Function const&, net::URLRequest const*, v8::Local<v8::Value> >(v8::Isolate* const&&&, mate::internal::SafeV8Function const&&&, net::URLRequest const*&&, v8::Local<v8::Value>&&) + 25
7   com.github.electron.framework 	0x00000001042fd6c5 void base::internal::InvokeHelper<false, void, base::internal::RunnableAdapter<void (*)(v8::Isolate*, mate::internal::SafeV8Function const&, net::URLRequest const*, v8::Local<v8::Value>)> >::MakeItSo<v8::Isolate* const&, mate::internal::SafeV8Function const&, net::URLRequest const*, v8::Local<v8::Value> >(base::internal::RunnableAdapter<void (*)(v8::Isolate*, mate::internal::SafeV8Function const&, net::URLRequest const*, v8::Local<v8::Value>)>, v8::Isolate* const&&&, mate::internal::SafeV8Function const&&&, net::URLRequest const*&&, v8::Local<v8::Value>&&) + 21
8   com.github.electron.framework 	0x00000001042fd697 base::internal::Invoker<base::IndexSequence<0ul, 1ul>, base::internal::BindState<base::internal::RunnableAdapter<void (*)(v8::Isolate*, mate::internal::SafeV8Function const&, net::URLRequest const*, v8::Local<v8::Value>)>, void (v8::Isolate*, mate::internal::SafeV8Function const&, net::URLRequest const*, v8::Local<v8::Value>), v8::Isolate*&, mate::internal::SafeV8Function>, base::internal::InvokeHelper<false, void, base::internal::RunnableAdapter<void (*)(v8::Isolate*, mate::internal::SafeV8Function const&, net::URLRequest const*, v8::Local<v8::Value>)> >, void (net::URLRequest const*, v8::Local<v8::Value>)>::Run(base::internal::BindStateBase*, net::URLRequest const*&&, v8::Local<v8::Value>&&) + 71
9   com.github.electron.framework 	0x00000001043659b7 base::Callback<void (net::URLRequest const*, v8::Local<v8::Value>), (base::internal::CopyMode)1>::Run(net::URLRequest const*, v8::Local<v8::Value>) const + 39
10  com.github.electron.framework 	0x0000000104365951 atom::internal::AskForOptions(v8::Isolate*, base::Callback<void (net::URLRequest const*, v8::Local<v8::Value>), (base::internal::CopyMode)1> const&, net::URLRequest*, base::Callback<void (v8::Isolate*, v8::Local<v8::Value>), (base::internal::CopyMode)1> const&, base::Callback<void (bool, std::__1::unique_ptr<base::Value, std::__1::default_delete<base::Value> >), (base::internal::CopyMode)1> const&) + 129
11  com.github.electron.framework 	0x000000010436a3cb void base::internal::RunnableAdapter<void (*)(v8::Isolate*, base::Callback<void (net::URLRequest const*, v8::Local<v8::Value>), (base::internal::CopyMode)1> const&, net::URLRequest*, base::Callback<void (v8::Isolate*, v8::Local<v8::Value>), (base::internal::CopyMode)1> const&, base::Callback<void (bool, std::__1::unique_ptr<base::Value, std::__1::default_delete<base::Value> >), (base::internal::CopyMode)1> const&)>::Run<v8::Isolate* const&, base::Callback<void (net::URLRequest const*, v8::Local<v8::Value>), (base::internal::CopyMode)1> const&, net::URLRequest* const&, base::Callback<void (v8::Isolate*, v8::Local<v8::Value>), (base::internal::CopyMode)1> const&, base::Callback<void (bool, std::__1::unique_ptr<base::Value, std::__1::default_delete<base::Value> >), (base::internal::CopyMode)1> const&>(v8::Isolate* const&&&, base::Callback<void (net::URLRequest const*, v8::Local<v8::Value>), (base::internal::CopyMode)1> const&&&, net::URLRequest* const&&&, base::Callback<void (v8::Isolate*, v8::Local<v8::Value>), (base::internal::CopyMode)1> const&&&, base::Callback<void (bool, std::__1::unique_ptr<base::Value, std::__1::default_delete<base::Value> >), (base::internal::CopyMode)1> const&&&) + 27
12  com.github.electron.framework 	0x000000010436a375 void base::internal::InvokeHelper<false, void, base::internal::RunnableAdapter<void (*)(v8::Isolate*, base::Callback<void (net::URLRequest const*, v8::Local<v8::Value>), (base::internal::CopyMode)1> const&, net::URLRequest*, base::Callback<void (v8::Isolate*, v8::Local<v8::Value>), (base::internal::CopyMode)1> const&, base::Callback<void (bool, std::__1::unique_ptr<base::Value, 
```

```
Thread 0 Crashed:: CrBrowserMain  Dispatch queue: com.apple.main-thread
0   com.github.electron.framework 	0x000000010185da36 atom::GetUploadData(base::ListValue*, net::URLRequest const*) + 38
1   com.github.electron.framework 	0x000000010185d906 mate::Converter<net::URLRequest const*, void>::ToV8(v8::Isolate*, net::URLRequest const*) + 390
2   com.github.electron.framework 	0x0000000101783b6c v8::Local<v8::Value> mate::ConvertToV8<net::URLRequest const*>(v8::Isolate*, net::URLRequest const* const&) + 12
3   com.github.electron.framework 	0x00000001017b345c mate::internal::V8FunctionInvoker<void (net::URLRequest const*, v8::Local<v8::Value>)>::Go(v8::Isolate*, mate::internal::SafeV8Function const&, net::URLRequest const*, v8::Local<v8::Value>) + 172
4   com.github.electron.framework 	0x00000001017b36e9 void base::internal::RunnableAdapter<void (*)(v8::Isolate*, mate::internal::SafeV8Function const&, net::URLRequest const*, v8::Local<v8::Value>)>::Run<v8::Isolate* const&, mate::internal::SafeV8Function const&, net::URLRequest const*, v8::Local<v8::Value> >(v8::Isolate* const&&&, mate::internal::SafeV8Function const&&&, net::URLRequest const*&&, v8::Local<v8::Value>&&) + 25
5   com.github.electron.framework 	0x00000001017b36c5 void base::internal::InvokeHelper<false, void, base::internal::RunnableAdapter<void (*)(v8::Isolate*, mate::internal::SafeV8Function const&, net::URLRequest const*, v8::Local<v8::Value>)> >::MakeItSo<v8::Isolate* const&, mate::internal::SafeV8Function const&, net::URLRequest const*, v8::Local<v8::Value> >(base::internal::RunnableAdapter<void (*)(v8::Isolate*, mate::internal::SafeV8Function const&, net::URLRequest const*, v8::Local<v8::Value>)>, v8::Isolate* const&&&, mate::internal::SafeV8Function const&&&, net::URLRequest const*&&, v8::Local<v8::Value>&&) + 21
```

### PR DESCRIPTION
**Pull request by deepak1556**
Fixes #5896 
Fixes #5925 
Fixes #4991 

What is the take on extensions and partitions ? Currently extensions are loaded for all windows irrespective of partitions, should extensions be persisted for non persistent partitions ?

### CODE REVIEW

File: spec/api-browser-window-spec.js [hunk 1]
```diff
@@ -879,6 +880,47 @@ describe('browser-window module', function () {
           })
         })
       })
+
+      describe('does not crash when a partition is used', function (done) {
```
Comment (kevinsawicki) [inline spec/api-browser-window-spec.js]: I don't think you need the `done` parameter on the `describe` block here.

**Comment by zcbenz** [comment]
This is mostly a work around of the original bug, I prefer to fix the root problem with following steps:

1. Fix the crash in #5896, so instead of crashing, the devtools should just be simply not working
1. Make the protocol module work with partitions
1. Add event for session creations for session module
1. Create the handler for chrome-extension protocol when a new session is created

**Comment by deepak1556** [comment]
Have made the changes, might have to add a few more tests. Thanks!

File: docs/api/session.md [hunk 2]
```diff
@@ -547,3 +547,24 @@ The `listener` will be called with `listener(details)` when an error occurs.
...
   * `fromCache` Boolean
   * `error` String - The error description.
+
+#### `ses.protocol`
+
+Returns an instance of [protocol](protocol.md) module for this session.
+
+```javascript
+const electron = require('electron');
+const app = electron.app;
```
Comment (zcbenz) [inline docs/api/session.md]: We can use destructuring assignment here:

```js
const {app, session} = require('electron')
```

And for new code samples in docs we should use new coding style, e.g. without `;`.

File: docs/api/session.md [hunk 3]
```diff
@@ -547,3 +547,24 @@ The `listener` will be called with `listener(details)` when an error occurs.
...
+const session = electron.session;
+const path = require('path');
+
+app.on('ready', function() {
```
Comment (zcbenz) [inline docs/api/session.md]: There should be a space after `function`.

File: lib/browser/api/session.js [hunk 4]
```diff
@@ -35,6 +35,7 @@ Object.defineProperty(exports, 'defaultSession', {
 const wrapSession = function (session) {
   // Session is an EventEmitter.
   Object.setPrototypeOf(session, EventEmitter.prototype)
+  process.emit('session-created', session)
```
Comment (zcbenz) [inline lib/browser/api/session.js]: How about making this event part of `session` module, so users can listen to it by:

```js
const {session} = require('electron')
session.on('session-created', () => {
})
```

File: spec/api-session-spec.js [hunk 5]
```diff
@@ -262,4 +272,50 @@ describe('session module', function () {
...
+        show: false,
+        width: 400,
+        height: 400,
+        webPreferences: {
+          partition: partitionName
+        }
+      })
+    })
+
+    afterEach(function () {
```
Comment (zcbenz) [inline spec/api-session-spec.js]: This seems redundant.

**Comment by deepak1556** [comment]
@zcbenz Fixed, thanks!

**Comment by zcbenz** [comment]
👍

**Pull request merged**